### PR TITLE
#262 ChronoZonedDateTime.isAfterOrEquals assertion

### PR DESCRIPTION
**Pull request by lukebiddell**
A lot of this is based on work done in #264.  Do I need to do the same for `ChronoLocalDateTime<D>` and `ChronoZonedDateTime<D>`?

----
I confirm that I have read the [Contributor Agreements v1.0](https://github.com/robstoll/atrium/blob/master/.github/Contributor%20Agreements%20v1.0.txt), agree to be bound on them and confirm that my contribution is compliant.

### CODE REVIEW

**Comment by robstoll** [comment]
Haven't looked at the PR yet but yes, you need 
 to do the same for ChronoLocalDateTime<D> and ChronoZonedDateTime<D>

File: translations/de_CH/atrium-translations-de_CH-common/src/main/kotlin/ch/tutteli/atrium/translations/DescriptionDateTimeLikeAssertion.kt [hunk 1]
```diff
@@ -11,5 +11,6 @@ enum class DescriptionDateTimeLikeAssertion(override val value: String) : String
     YEAR("Jahr"),
     MONTH("Monat"),
     DAY_OF_WEEK("Wochentag"),
-    DAY("Tag")
+    DAY("Tag"),
+    IS_AFTER_OR_EQUALS("ist nach dem oder gleich"),
```
Comment (robstoll) [inline translations/de_CH/atrium-translations-de_CH-common/src/main/kotlin/ch/tutteli/atrium/translations/DescriptionDateTimeLikeAssertion.kt]: change to 
```suggestion
    IS_AFTER_OR_EQUALS("ist am oder nach dem"),
```

**Comment by lukebiddell** [comment]
Can't seem to get the Spec tests working. In the file `kotlin/ch/tutteli/atrium/api/fluent/en_GB/ChronoLocalDateAssertionsSpec.kt`, isAfterOrEquals cannot be resolved.

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/ChronoLocalDateAssertionsSpec.kt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package ch.tutteli.atrium.api.fluent.en_GB
+
+import ch.tutteli.atrium.creating.Expect
+import ch.tutteli.atrium.specs.fun1
+
+object ChronoLocalDateAssertionsSpec : ch.tutteli.atrium.specs.integration.ChronoLocalDateAssertionsSpec(
+    fun1(Expect<Int>::isAfterOrEquals) // Unresolved reference: isAfterOrEquals
```
Comment (robstoll) [inline apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/ChronoLocalDateAssertionsSpec.kt]: isAfterOrEquals does not exist for `Expect<Int>` you need to use `Expect<ChronoLocalDate>`

File: misc/specs/atrium-specs-common/src/main/kotlin/ch/tutteli/atrium/specs/integration/ChronoLocalDateAssertionsSpec.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import ch.tutteli.atrium.specs.*
+import ch.tutteli.atrium.translations.DescriptionDateTimeLikeAssertion
+import org.spekframework.spek2.Spek
+import org.spekframework.spek2.style.specification.describe
+
+abstract class ChronoLocalDateAssertionsSpec(
+//    isBefore: Fun1<Int, Int>,
+//    isBeforeOrEquals: Fun1<Int, Int>,
+//    isAfter: Fun1<Int, Int>,
+    isAfterOrEquals: Fun1<Int, Int>,
```
Comment (robstoll) [inline misc/specs/atrium-specs-common/src/main/kotlin/ch/tutteli/atrium/specs/integration/ChronoLocalDateAssertionsSpec.kt]: same here, you have to work with ChronoLocalDate and not with Int

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/ChronoLocalDateAssertionsSpec.kt (hunk 2 above)
Comment (lukebiddell) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/ChronoLocalDateAssertionsSpec.kt]: Ah that makes a lot of sense now that you mention it. Strangely however ChronoLocalDate is unresolved too. 😞

Comment (robstoll) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/ChronoLocalDateAssertionsSpec.kt]: give me a sec, I will checkout your branch and check

Comment (robstoll) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/ChronoLocalDateAssertionsSpec.kt]: ah... you have placed the file in the wrong project. You need to move it to the atrium-api-fluent-en_GB-jdk8-common

Comment (lukebiddell) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/ChronoLocalDateAssertionsSpec.kt]: That package doesn't contain anything? The build.gradle says "dummy project, will be removed once Atrium switches to the new MPP plugin."

Comment (robstoll) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/ChronoLocalDateAssertionsSpec.kt]: my fault, move it to atrium-api-fluent-en_GB-jdk8-jvm next to your implementation basically. This way you see your implementation in the test

File: misc/specs/atrium-specs-jvm/src/main/kotlin/ch/tutteli/atrium/specs/integration/ChronoLocalDateAssertionsSpec.kt [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+    val isAfterOrEqualsDescr = DescriptionDateTimeLikeAssertion.IS_AFTER_OR_EQUALS.getDefault()
+
+    val fluent = expect(LocalDate.of(2019, 12, 15) as ChronoLocalDate)
+
+    describe("$describePrefix context subject is 2019/12/15") {
+        describe("${isAfterOrEquals.name} ...") {
+            val isAfterOrEqualsFun = isAfterOrEquals.lambda
+
+            it("... 2019/12/16 throws an AssertionError containing ${DescriptionDateTimeLikeAssertion::class.simpleName}.${DescriptionDateTimeLikeAssertion.IS_AFTER_OR_EQUALS} and `: 11`") {
```
Comment (robstoll) [inline misc/specs/atrium-specs-jvm/src/main/kotlin/ch/tutteli/atrium/specs/integration/ChronoLocalDateAssertionsSpec.kt]: IMO good enough if you write only
```suggestion
            it("... 2019/12/16 throws an AssertionError") {
```
if you want to keep it this way, then please fix `: 11`, should be `: 2019/12/16`

**Comment by lukebiddell** [comment]
Still having a few problems. I'm not sure `object ChronoLocalDateAssertionsSpec` is in the correct package, as isAfterOrEquals is an unresolved reference (IntelliJ offers to import it and adds it as an import, but it doesn't work).

Secondly, with `abstract class ChronoLocalDateAssertionsSpec`, I'm having issues with `forSubjectLess` on line 22 (in fact I'm not sure what SubjectLessSpec is for, or whether I should be using it).

Sorry for the questions, I haven't dealt with a large code base like this before.

File: apis/fluent-en_GB/extensions/jdk8/atrium-api-fluent-en_GB-jdk8-jvm/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/jdk8/ChronoLocalDateAssertionsSpec.kt [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package ch.tutteli.atrium.api.fluent.en_GB.jdk8
+
+import ch.tutteli.atrium.creating.Expect
+import ch.tutteli.atrium.domain.builders.creating.ChronoLocalDateAssertionsBuilder.isAfterOrEquals
+import ch.tutteli.atrium.specs.fun1
+import java.time.chrono.ChronoLocalDate
+
+object ChronoLocalDateAssertionsSpec : ch.tutteli.atrium.specs.integration.ChronoLocalDateAssertionsSpec(
+    fun1(Expect<ChronoLocalDate>::isAfterOrEquals) // Unresolved reference: isAfterOrEquals
```
Comment (robstoll) [inline apis/fluent-en_GB/extensions/jdk8/atrium-api-fluent-en_GB-jdk8-jvm/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/jdk8/ChronoLocalDateAssertionsSpec.kt]: First of all, no worries, it's a good first issue I am not expecting that you see through the code at this stage.
You need to add an isAfterOrEquals method to the API as well (I thought I saw it before, have you removed it again?). Consult the to-do list in the issue description to check if other things are missing.

File: misc/specs/atrium-specs-jvm/src/main/kotlin/ch/tutteli/atrium/specs/integration/ChronoLocalDateAssertionsSpec.kt [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+abstract class ChronoLocalDateAssertionsSpec(
+    //isBefore, isBeforeEquals, isAfter
+    isAfterOrEquals: Fun1<ChronoLocalDate, ChronoLocalDate>,
+    describePrefix: String = "[Atrium] "
+) : Spek({
+
+    include(object : SubjectLessSpec<ChronoLocalDate>(
+        describePrefix,
+        isAfterOrEquals.forSubjectLess(1)
```
Comment (robstoll) [inline misc/specs/atrium-specs-jvm/src/main/kotlin/ch/tutteli/atrium/specs/integration/ChronoLocalDateAssertionsSpec.kt]: Your isAfterOrEquals method expects a ChronoLocalDate and not an Int

**Comment by robstoll** [comment]
@lukebiddell please rebase on master and have a look at https://github.com/robstoll/atrium/pull/282/files#diff-1f59c5fff9384d427b3573e5b3a93fc9R13 for the specs

File: translations/de_CH/atrium-translations-de_CH-common/src/main/kotlin/ch/tutteli/atrium/translations/DescriptionDateTimeLikeAssertion.kt [hunk 7]
```diff
@@ -11,5 +11,6 @@ enum class DescriptionDateTimeLikeAssertion(override val value: String) : String
     YEAR("Jahr"),
     MONTH("Monat"),
     DAY_OF_WEEK("Wochentag"),
-    DAY("Tag")
+    DAY("Tag"),
+    IS_AFTER_OR_EQUALS("ist am oder nach dem"),
```
Comment (robstoll) [inline translations/de_CH/atrium-translations-de_CH-common/src/main/kotlin/ch/tutteli/atrium/translations/DescriptionDateTimeLikeAssertion.kt]: I realised the text I proposed only works well for ChronoLocalDate, please change to:
```suggestion
    IS_AFTER_OR_EQUALS("ist nach oder zeitgleich"),
```

**Comment by lukebiddell** [comment]
I believe it's done now, let me know if any changes are required.

File: misc/specs/atrium-specs-jvm/src/main/kotlin/ch/tutteli/atrium/specs/integration/ChronoLocalDateAssertionSpec.kt [hunk 8]
```diff
@@ -89,6 +93,21 @@ abstract class ChronoLocalDateAssertionSpec(
                     }.toThrow<AssertionError> { messageContains("$isAfterDescr: $december24") }
                 }
             }
+            describe("${isBeforeOrEquals.name} ...") {
```
Comment (robstoll) [inline misc/specs/atrium-specs-jvm/src/main/kotlin/ch/tutteli/atrium/specs/integration/ChronoLocalDateAssertionSpec.kt]:
```suggestion
            describe("${isAfterOrEquals.name} ...") {
```
But I'll fix it myself

**Comment by robstoll** [comment]
@lukebiddell thank you very much for your contribution :+1:

**Pull request merged**